enhancements/compound-bundle-constraints.md

```diff
@@ -0,0 +1,350 @@
+        group: foos.example.com
+        version: v1
+        kind: Foo
+  - type: olm.all.required
+    value:
+    - type: olm.package.required
+      value:
+        packageName: foo
+        versionRange: '<1.0.0'
+    - type: olm.gvk.required
+      value:
+        group: foos.example.com
+        version: v1beta1
+        kind: Foo
+```
+
+#### Changes to OLM's resolver
+
+The [resolver package][olm-resolver-primitives] supplies the primitives "and" and "or" already,
+and uses them to combine simple GVK and package constraints into an expression.
+Compound constraints would simply extend the usage of these primitives in the nested
+case, and add the "not" primitive.
+
+### Risks and Mitigations
+
+- With the possibility for nesting comes the increased possibility that two or more nestings
+will logically conflict and result in an unsatisfiable expression.
+  - There is no quick way to determine whether an expression is unsatisfiable
+  other than by using a SAT solver, which the resolver uses internally.
+  Compound constraints must be vetted by hand.
+- Compound constraint nesting exposes an exponentiation attack vector, where a malicious
+bundle author adds deep nesting to a bundle resulting in long resolver pauses.
+  - Tools that work with bundle constraints should limit the depth of nesting.
+  I recommend the depth limit be 10 levels.
+  - Typically bundles are vetted by catalog maintainers, so deep nesting can be caught
+  by automation prior to inclusion in a catalog.
+
+## Design Details
+
+### Test Plan
+
+- operator-registry unit and e2e tests.
+- operator-lifecycle-manager unit and e2e tests.
+  - Both API and resolver unit tests.
+
+### Graduation Criteria
+
+The alpha feature gate will be removed and this feature will be on permanently by OLM v1.0.0.
+
+#### Examples
+
+See ["Implementation Details/Notes/Constraints"](#implementation-details-notes-constraints).
+
+##### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+
+##### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
+##### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+Older versions of OLM will ignore unknown bundle properties, so compound constraints
+will be ignored. Starting with the first release containing this feature, the same
+holds for all properties not listed as feature gated properties; if the feature gate
+is not turned on for some property, OLM will emit an error. Therefore the same bundle
+will "work" between OLM versions, even though older versions will install the bundle
+without error by design.
```

Comment:
Out-of-scope for this particular EP, but somewhat foundational. If we know we'll eventually want to support `optional` and/or `recommended` constraints, then it seems like it may be better to make a separate top-level constraint field for this enum value. Something like:
```yaml
type: olm.all
necessity: required|recommended|optional
value: { ... }
```

Reply:
Moved to version skew. I agree, distinguishing between constraint necessity would be valuable. Worthy of an operator-registry GH issue?